components/dht/dht_memory.go

```diff
@@ -44,128 +22,107 @@ func Create() *dht {
-
-	d := &dht{
-		slotVsNodes: make(map[SlotID]*SlotInfo),
-	}
-	// nodeVsSlot contains the mapping of nodeID to slotID.
-	nodeVsSlot := make(map[NodeID][]SlotID)
-
-	nodeCount := len(nodes)
-	slotCount := slotCountperNode * nodeCount
+func InitialiseDHT(shardCount int, seedNodes []string, replication int) (map[ShardID]ShardLocation, error) {
+	shards := make(map[ShardID]ShardLocation, shardCount)
+	nodeCount := len(seedNodes)
 
-	// The distribution below makes sure the slots are
-	// assigned equally in a round robin manner
-	distribution := make([]int, nodeCount)
-	for i := 0; i < slotCount; i++ {
-		distribution[i%nodeCount]++
+	if nodeCount < replication {
+		return nil, ErrReplicasLessThanNodes
 	}
 
-	slotNumber := 0
-	for i := 0; i < len(distribution); i++ { // For each of the node
-		for j := 0; j < distribution[i]; j++ { // For the distribution count assigned to that node
-			nodeID := NodeID(nodes[i])
-			slotID := SlotID(slotNumber)
+	for i := 0; i < shardCount; i++ {
+		leaderNode := NodeID(seedNodes[i%nodeCount])
+		followerNodes := []NodeID{}
 
-			d.slotVsNodes[slotID] = &SlotInfo{
-				NodeID: nodeID,
-			}
-			nodeVsSlot[nodeID] = append(nodeVsSlot[nodeID], slotID)
-			slotNumber++
+		for r := 1; r < replication; r++ {
+			follower := NodeID(seedNodes[(i+r)%nodeCount])
+			followerNodes = append(followerNodes, follower)
 		}
-	}
-
-	// Assign leaders in a round robin manner
-	for i := 0; i < slotCountperNode; i++ {
-		for nodeID := range nodeVsSlot {
-			slotID := nodeVsSlot[nodeID][i]
-			slotInfo := d.slotVsNodes[slotID]
-			if slotInfo.SlotState != "" {
-				continue
-			}
-			slotInfo.SlotState = Leader
 
-			replicaSlotInfo := d.slotVsNodes[d.replicaSlot(slotID)]
-			replicaSlotInfo.SlotState = Follower
+		shardID := ShardID(i)
+		shards[shardID] = ShardLocation{
+			ID:        shardID,
+			Leader:    leaderNode,
+			Followers: followerNodes,
 		}
 	}
 
-	by, _ := json.Marshal(d.slotVsNodes) // TODO: Remove
-	fmt.Printf("Slot Info: %s\n", string(by))
-
-	return d.slotVsNodes, nil
+	return shards, nil
 }
 
-// Load loads the slot info data from a already existing configuration to a given empty dht.
-// This must be called only after confirmation from the master.
-func (d *dht) Load(slots map[SlotID]*SlotInfo) error {
-	d.mu.Lock()
-	defer d.mu.Unlock()
+func (d *dht) GetShard(key string) (ShardLocation, error) {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
 
-	d.slotVsNodes = slots // TODO: Verify if this will share memory
-	return nil
+	return d.shards[d.hashSlot(key)], nil
 }
 
-// Snapshot returns the node vs slot ids map in json format
-func (d *dht) Snapshot() map[SlotID]*SlotInfo {
-	d.mu.Lock()
-	defer d.mu.Unlock()
-
-	return d.slotVsNodes // TODO: Verify if this will share memory
+func (d *dht) hashSlot(key string) ShardID {
+	slotCount := uint64(len(d.shards))
+	hashValue := xxhash.Sum64([]byte(key))
+	return ShardID(hashValue % slotCount)
```

Comment:
Since shardID is dependent on shard count while increasing/decreasing shards, we would corrupt the existing shardIDs, we need a stable algorithm to calculate shard IDs or some other mechanism.

Reply:
Yes, once the shard count is configured, you cant change it.  If they still want to do it, then it will require migration.

To start off, we can set the inital shards to 256. We can scale this upto 256 nodes in the future which is very unlikely.